client/client.go

```diff
@@ -431,14 +437,24 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
 		// Check 2xx status codes
 		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
 			obj, err := container.ParseJSON(bodyBytes)
-			if err == nil {
-				log.Printf("[DEBUG] Exit from do method")
-				return obj, resp, err
+			if err != nil {
+				// Attempt retry if JSON parsing fails but status code is 2xx
+				// Assumption here is that packets were somehow corrupted/lost during transmission
+				log.Printf("[ERROR] Error occurred while JSON parsing (2xx status): %+v", err)
+				retry = true
+			} else {
+				// JSON parsing was successful for a 2xx response.
+				// Now, check the custom retry function.
+				if retryFunc != nil && retryFunc(obj) {
+					log.Printf("[DEBUG] Custom retry function indicated a retry is needed for 2xx response")
+					retry = true
+				} else {
+					// If JSON parsed successfully and retryFunc does not indicate a retry,
+					// then this is a successful operation.
+					log.Printf("[DEBUG] Exit from Do method (2xx success, no custom retry)")
```

Comment:
Maybe we should log `no custom retry function` only when retryFunc == nil and `with custom retry function` when its not nil? Just so we know during debugging that a retry function is used or not. Currently the only way to know its is definitely used in debugging is if the function returns false and logs the `Custom retry function indicated...` message.

Reply:
I removed the unnecessary string in the log. Is this ok?